src/Integration.Vsix.UnitTests/Helpers/ActiveDocumentTrackerTests.cs

```diff
@@ -0,0 +1,209 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using FluentAssertions;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Microsoft.VisualStudio.Text;
+using Moq;
+using SonarLint.VisualStudio.Integration.Vsix;
+using SonarLint.VisualStudio.Integration.Vsix.Helpers.DocumentEvents;
+
+namespace SonarLint.VisualStudio.Integration.UnitTests.Helpers
+{
+    [TestClass]
+    public class ActiveDocumentTrackerTests
+    {
+        private Mock<IVsMonitorSelection> monitorSelectionMock;
+        private Mock<IServiceProvider> serviceProviderMock;
+        private Mock<Action<DocumentFocusedEventArgs>> mockEventHandler;
+        private Mock<ITextDocumentProvider> textDocumentProviderMock;
+        private ITextDocument textDocumentMock;
+
+        private ActiveDocumentTracker testSubject;
+
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            ThreadHelper.SetCurrentThreadAsUIThread();
+
+            textDocumentMock = Mock.Of<ITextDocument>();
+
+            monitorSelectionMock = new Mock<IVsMonitorSelection>();
+
+            serviceProviderMock = new Mock<IServiceProvider>();
+            serviceProviderMock
+                .Setup(x => x.GetService(typeof(SVsShellMonitorSelection)))
+                .Returns(monitorSelectionMock.Object);
+
+            textDocumentProviderMock = new Mock<ITextDocumentProvider>();
+            mockEventHandler = new Mock<Action<DocumentFocusedEventArgs>>();
+
+            testSubject = new ActiveDocumentTracker(serviceProviderMock.Object, textDocumentProviderMock.Object);
+            testSubject.OnDocumentFocused += (sender, e) => mockEventHandler.Object(e);
+        }
+
+        [TestMethod]
+        public void Ctor_ShouldRegisterToSelectionEvents()
+        {
+            uint cookie = 1234;
+            monitorSelectionMock.Reset();
```

Comment:
Is this `Reset` required? Nothing has happened since initialization.

Reply:
Sort of: I init monitorSelectionMock in the TestInitialize, which is then passed to another instance of ActiveDocumentTracker, which calls AdviseSelectionEvents on it.
So then when I call `monitorSelectionMock.VerifyNoOtherCalls` - it throws that there was an additional invocation.

Reply:
Ah. That illustrates what's wrong with [TestInitialize] unless it applies to every test, which it almost never does.

It's not obvious why the reset is there, so at least it deserves a comment.